In [1]:
import tensorflow as tf

In [2]:
mnist = tf.keras.datasets.mnist
(x_train, y_train), (x_test, y_test) = mnist.load_data()

In [3]:
x_train, x_test = x_train/255.0, x_test/255.0

In [4]:
x_train = x_train.reshape(60000, 28*28)
x_test = x_test.reshape(10000, 28*28)

In [11]:
model = tf.keras.models.Sequential([
    tf.keras.layers.Dense(16, activation='tanh', input_shape=(784,)),
    tf.keras.layers.Dense(10)
])

In [6]:
predictions = model(x_train[:1])
predictions = predictions.numpy()

In [7]:
tf.nn.softmax(predictions).numpy()

array([[0.12038827, 0.05729026, 0.0949445 , 0.10288949, 0.12614827,
        0.09920219, 0.10036646, 0.09418786, 0.1030702 , 0.10151245]],
      dtype=float32)

In [8]:
loss_fn = tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True)
loss_fn(y_train[:1], predictions).numpy()

2.310595

In [14]:
model.compile(optimizer='sgd', loss=loss_fn, metrics=['accuracy'])
model.fit(x_train, y_train, epochs=5)

Epoch 1/5
1875/1875 [==============================] - 3s 2ms/step - loss: 0.1392 - accuracy: 0.9606
Epoch 2/5
1875/1875 [==============================] - 3s 1ms/step - loss: 0.1366 - accuracy: 0.9616
Epoch 3/5
1875/1875 [==============================] - 3s 1ms/step - loss: 0.1352 - accuracy: 0.9622
Epoch 4/5
1875/1875 [==============================] - 3s 1ms/step - loss: 0.1343 - accuracy: 0.9627
Epoch 5/5
1875/1875 [==============================] - 3s 1ms/step - loss: 0.1334 - accuracy: 0.9625


In [10]:
model.evaluate(x_test, y_test, verbose=2)

313/313 - 1s - loss: 0.1949 - accuracy: 0.9446 - 549ms/epoch - 2ms/step


[0.19491462409496307, 0.944599986076355]

## For the options permutation
### At the first run
    Option used => adam and tanh
    train time => 16s
    train accuracy => 0.9486
    test accuracy => 0.9446
    
### At the second run
    Option used => rmspropo and tanh
    train time => 16s
    train accuracy => 0.9560
    test accuracy => 0.9446

### At the thrid run
    Option used => rmspropo and tanh
    train time => 15s
    train accuracy => 0.9625
    test accuracy => 0.9446